src/WeakMap.php

```diff
@@ -191,5 +202,41 @@ private function assertValidKey($key) : void
                 throw new TypeError('WeakMap key must be an object');
             }
         }
+
+        private function setupHousekeepingOnGcRun() : void
+        {
+            $this->destructedEarly = new \stdClass();
+
+            $weakThis = \WeakReference::create($this); // allow self to be released without GC
+            $weakThisDestructedEarly = \WeakReference::create($this->destructedEarly);
+
+            $gcRuns = 0;
+            $setupDestructorFx = static function () use ($weakThis, $weakThisDestructedEarly, &$gcRuns, &$setupDestructorFx): void {
+                $destructorFx = static function () use ($weakThis, $weakThisDestructedEarly, &$gcRuns, &$setupDestructorFx): void {
+                    $gcRunsPrev = $gcRuns;
+                    $gcRuns = gc_status()['runs'];
+                    if ($gcRunsPrev !== $gcRuns) { // prevent recursion on shutdown
+                        $setupDestructorFx();
+                    }
+
+                    if ($weakThisDestructedEarly->get() === null) {
+                        return;
+                    }
+
+                    $map = $weakThis->get();
+                    if ($map !== null) {
+                        \Closure::bind(static fn () => $map->performHousekeepingOnGcRun(), null, \WeakMap::class)();
```

Comment:
Is the private method required? Can't it be `static fn () => $map->housekeeping(true)`?

Reply:
good catch!